scripts/build.py

```diff
@@ -715,62 +874,80 @@ def generate_build_directory():
+        write_to_file_stream(
+            hashes_js_file, get_hashes_json_file_contents(hashes))
+    # Update hash dict with newly created hashes.js.
+    hashes.update({HASHES_JS_FILENAME: generate_md5_hash(HASHES_JSON)})
+    # Make sure hashes.js is available to the frontend.
+    _ensure_files_exist([HASHES_JSON])
+
+    # Build files in /extensions and copy them into staging directory.
+    build_directory(EXTENSIONS_DIR, hashes, build_tasks)
     # Minify all template files copy them into build/templates/head.
-    if not os.path.isdir(TEMPLATES_STAGING_DIR):
-        # If there is no staging dir, perform build process on all files.
-        print 'Creating new %s folder' % TEMPLATES_STAGING_DIR
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes)
-    else:
-        # If there is an existing staging dir, rebuild all HTML files.
-        print 'Staging directory exists, re-building all HTML files'
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes,
-            file_formats=('.html',))
-        recently_changed_filenames = get_recently_changed_filenames(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_OUT_DIR)
-        if recently_changed_filenames:
-            # Only re-build files that have changed since last build.
-            print ("Re-building recently changed files at %s"
-                   % TEMPLATES_DEV_DIR_CORE)
-            rebuild_new_files(
-                TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR,
-                recently_changed_filenames, hashes)
-            # Clean up files in staging dir that have been removed in DEV dir.
-            remove_deleted_files(TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR)
-        else:
-            print 'No changes detected. Using previously built files.'
-    # Copy built files from staging dir to build dir.
-    copy_files_source_to_target(
-        TEMPLATES_STAGING_DIR, TEMPLATES_OUT_DIR, hashes)
+    build_directory(TEMPLATES_CORE_DIR, hashes, build_tasks)
+    # Execute all build tasks.
+    try:
+        _execute_tasks(build_tasks)
+    except Exception as e:
+        print e
+
+    # Copy alls files from staging directory to production directory (/build).
+    COPY_INPUT_DIRS = [
+        ASSETS_DEV_DIR, EXTENSIONS_DIR.get('staging_dir'),
+        TEMPLATES_CORE_DIR.get('staging_dir'),
+        THIRD_PARTY_GENERATED_STAGING_DIR]
+    COPY_OUTPUT_DIRS = [
+        ASSETS_OUT_DIR, EXTENSIONS_DIR.get('out_dir'),
+        TEMPLATES_CORE_DIR.get('out_dir'), THIRD_PARTY_GENERATED_OUT_DIR]
+    assert len(COPY_INPUT_DIRS) == len(COPY_OUTPUT_DIRS)
+    for i in xrange(len(COPY_INPUT_DIRS)):
+        copy_files_source_to_target(
+            COPY_INPUT_DIRS[i], COPY_OUTPUT_DIRS[i], hashes, copy_tasks)
+    # Execute all copy tasks.
+    try:
+        _execute_tasks(copy_tasks)
+    except Exception as e:
+        print e
+
+    # Rewrite hashes.js as JSON again due to deletion by copy functions.
```

Comment:
This sort of thing isn't great, FYI. Can you see if there's a way to make it cleaner? If a file is going to get deleted and rewritten again then perhaps we just shouldn't write it the first time.

I feel like our build process is getting too complicated, and any steps that can be taken to simplify the flow would be very helpful.

Reply:
Yep, I am now writing hashes.js to /assets and clean it up after. Thanks!